### PR TITLE
Loaders for LSDB and hipscat (reimplementation using LSDB) data

### LINKED ISSUE #360: Add Ensemble.load_lsdb
Currently we load an Ensemble from an LSDB catalogue by pulling out the `_ddf` object fields from LSDB's catalogue objects.

Instead we can provide a cleaner `Ensemble.load_lsdb` method as the default entry point for a user working from LSDB.

### PR DESCRIPTION
**Pull request by dougbrn**
## Change Description

Resolves #360. Adds an Ensemble loader function for LSDB catalog objects. Reimplements Ensemble.from_hipscat to use LSDB under the hood. Adds a new tutorial to the docs, potentially addressing #367 
- [x] My PR includes a link to the issue that I am addressing

## Solution Description

The main implementation for both functions is fairly straightforward, just piping the LSDB catalog dataframes into Ensemble.from_dask_dataframe. There's a few things that contributed to API/implementation complexity:

* Ensemble.from_hipscat needs the object_index and source_index kwargs in addition to the id_col specified in the column mapper. These are needed to do a join between object and source to reseat the source index on top of the object level _hipscat_index. I tried to add information to the docstring to inform on how these should be used.
* Support for no object catalog; in keeping with TAPE tradition, I wanted to keep the ability to load in from just a source table. This adds some complication to the implementation, as you should not use the _hipscat_index in this case (it's unique per source). And if you don't use the _hipscat_index, you now almost certainly need to sort. I tried to add a few warnings to catch some problematic inputs, and even added a try except to fall back to using sort=True in cases where the user provided incorrect sort/sorted kwargs. My rationale being that it's better for user experience to warn them of a kwarg mistake but still yield a valid result.
* Ability to choose an index other than _hipscat_index. When you load both an object and source catalog, I think it's pretty much always correct (?) to use the _hipscat_index. We should be pre-sorted by it, and it's guaranteed to be in both catalogs. However a user could choose something else in the column mapper API. I provided a longer warning in these cases about what this other id column will need, but opted not to force the user into using the _hipscat_index (although I'm still definitely considering it).

## Code Quality
- [x] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

## Project-Specific Pull Request Checklists
- [ ] I have added a function that requires a sync_tables command, and have added the neccesary sync_tables call

### Bug Fix Checklist
- [ ] My fix includes a new test that breaks as a result of the bug (if possible)
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### New Feature Checklist
- [x] I have added or updated the docstrings associated with my feature using the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [x] I have updated the tutorial to highlight my new feature (if appropriate)
- [x] I have added unit/End-to-End (E2E) test cases to cover my new feature
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### Documentation Change Checklist
- [ ] Any updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)

### Build/CI Change Checklist
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the README to reflect this
- [ ] If this is a new CI setup, I have added the associated badge to the README

### Other Change Checklist
- [ ] Any new or updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover any changes
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### CODE REVIEW

File: src/tape/ensemble_readers.py [hunk 1]
```diff
@@ -245,51 +245,130 @@ def read_parquet(
...
-        Path to the directory structure
-    source_subdir: 'str'
-        Path to the subdirectory which contains source files
-    object_subdir: 'str'
-        Path to the subdirectory which contains object files, if None then
-        files will only be read from the source_subdir
+    source_path: 'dask.Dataframe'
+        A hipscat directory that contains source information to be read
+        into the ensemble.
+    object_catalog: 'dask.Dataframe', optional
```
Comment (hombit) [inline src/tape/ensemble_readers.py:332]:
```suggestion
    object_path: str or Path, optional
```

File: src/tape/ensemble_readers.py [hunk 2]
```diff
@@ -245,51 +245,130 @@ def read_parquet(
...
     Parameters
     ----------
-    dir: 'str'
-        Path to the directory structure
-    source_subdir: 'str'
-        Path to the subdirectory which contains source files
-    object_subdir: 'str'
-        Path to the subdirectory which contains object files, if None then
-        files will only be read from the source_subdir
+    source_path: 'dask.Dataframe'
```
Comment (hombit) [inline src/tape/ensemble_readers.py:329]:
```suggestion    
    source_path: str or Path
```

**Review by hombit — COMMENTED** [review]
Thank you! It looks really good, I ask for just few small fixes

File: src/tape/ensemble.py [hunk 3]
```diff
@@ -1587,47 +1590,192 @@ def from_dask_dataframe(
...
+            joined_source_catalog = object_catalog.join(
+                source_catalog,
+                left_on=object_index,
+                right_on=source_index,
+                suffixes=("_drop_these_cols", ""),
+            )
+
+            # We can assume it's sorted if object and source are both present
+            # Though this depends on the chosen _id_col, as we assume
+            # _hipscat_index is going to be used
```
[the remark this answers is not in the document]
Comment (dougbrn) [reply src/tape/ensemble.py:1758]: I think we don't need to, as `from_lsdb` is handling a lot of the checking which will also be applied to the `from_hipscat` call. I'll be pushing a commit in a bit that should generally set the behavior that using an id col that is not _hipscat_index with the sorted flag true will raise a warning and set the sort/sorted kwargs to True/False respectively.

File: src/tape/ensemble.py [hunk 4]
```diff
@@ -1587,47 +1590,192 @@ def from_dask_dataframe(
...
+        if object_catalog is None:
+            # This is tricky, so just raise an explicit warning
+            if column_mapper.map["id_col"] == "_hipscat_index":
+                warnings.warn(
+                    "Using the _hipscat_index as the id column is not advised without a specified object catalog, as the _hipscat_index is unique per source in this case. Use an object-level id.",
+                    UserWarning,
+                )
+            # And if they didn't choose _hipscat_index, it's almost certainly not sorted
+            # Let's try to catch a bad sorted set, and reroute to sort for better user experience
+            try:
```
[the remark this answers is not in the document]
Comment (dougbrn) [reply src/tape/ensemble.py]: Changed this to just warn and reset the sort/sorted kwargs

File: src/tape/ensemble.py [hunk 5]
```diff
@@ -1587,47 +1590,192 @@ def from_dask_dataframe(
...
-        source_files = glob.glob(os.path.join(source_path, "**", "*.parquet"), recursive=True)
+        # Support for just source catalog is somewhat involved
+        # The code below mainly tries to catch a few common pitfalls
+        if object_catalog is None:
+            # This is tricky, so just raise an explicit warning
+            if column_mapper.map["id_col"] == "_hipscat_index":
+                warnings.warn(
+                    "Using the _hipscat_index as the id column is not advised without a specified object catalog, as the _hipscat_index is unique per source in this case. Use an object-level id.",
+                    UserWarning,
+                )
```
[the remark this answers is not in the document]
Comment (dougbrn) [reply src/tape/ensemble.py]: Done

**Review by hombit — APPROVED** [review]

**Pull request merged**